### ISSUE TITLE
Shape Mismatch Issues in Feature Aggregation

### ISSUE THREAD

**Issue by David-Lmh** [SCUT-BIP-Lab/PointDC#7]
I'm trying to reproduce your code for eval for scannet, and when using the torch_scatter library to aggregate features, I'm encountering runtime errors related to shape mismatches, as shown in the following problem:

![image](https://github.com/user-attachments/assets/77cb5d33-f65a-4646-8a0e-cdc58ff6ff29)

I ran into a similar problem while training and followed up with a break in training. I then used the experimental data you posted on Baidu.com （240523_default.zip） for eval and reported this error as well. The exact rows are shown below:

![image](https://github.com/user-attachments/assets/5b31ae48-3bd3-4059-bc6d-54b2c497de93)

For your code, I haven't made any changes and would like to ask if this reported error is something you've encountered/how it was resolved?

**Comment by HalvesChen** [comment]
Hi !
Could you show the `raw_region.shape inverse_map.shape ` and no inverse `feats_nonorm.shape` ？Looks like it's because the region(superpoint) isn't remapped to the point, change line127 to `region_feats = scatter(feats_norm, raw_region[inverse_map.long()].cuda(), dim=0, reduce='mean')` might work. If it works, please let me know~

**Comment by David-Lmh** [comment]
I tried your suggestion `region_feats = scatter(feats_norm, raw_region[inverse_map.long()].cuda(), dim=0, reduce='mean')` and found that it still doesn't work. Also, `raw_region.shape`, `inverse_map.shape` and non inverse `features_nonorm.shape` are shown below

![image](https://github.com/user-attachments/assets/db367185-598f-49d5-8159-cb7dd219a15b)

**Comment by HalvesChen** [comment]
Hi ! This is my modified code. Please try again.
```
coords, features, _, labels, inverse_map, pseudo_labels, inds, region, index, scenenames = data
region = region.squeeze()
raw_region = region.clone() # point region idx

in_field = ME.TensorField(features, coords, device=0)

if submodel is not None:
    feats_TensorField = model(in_field)
    feats_nonorm = submodel(feats_TensorField)
else:
    feats_nonorm = model(in_field)
feats_nonorm = feats_nonorm[inverse_map.long()] # point feats to voxel feats
feats_norm = F.normalize(feats_nonorm, dim=1) ## NOTE 可能需要normalize？没啥区别

region = region[inverse_map.long()] # point region idx to voxel region idx
region_feats = scatter(feats_norm, region.cuda(), dim=0, reduce='mean') # region feats

valid_mask = labels!=-1 # 获取带训练的mask区域
# labels = labels[valid_mask]
region_masked = raw_region[valid_mask].long()
region_masked_num = torch.unique(region_masked) # get no-masked region
region_masked_feats = region_feats[region_masked_num] # get no-masked region feats
region_masked_feats_norm = F.normalize(region_masked_feats, dim=1).cpu() # norm

region_feats_list.append(region_masked_feats_norm)

torch.cuda.empty_cache()
torch.cuda.synchronize(torch.device("cuda"))
```

**Comment by David-Lmh** [comment]
Same error, same shape reported, I'm wondering if it has something to do with a problem with my initial superpoint data. Can you please provide me a zip file of the superpoint data / or I can run the initial again to see the result.

**Comment by David-Lmh** [comment]
In the code for the initial sp scannet, I see a line that reads `parser.add_argument(‘--input_path’, type=str, default=‘data/ScanNet/scans/’, help=‘raw data path’) # *.segs .json file path`. Each of the `*.segs.json`s in the original dataset is one of two, e.g. ‘scene0000_00_vh_clean_2.0.010000.segs.json’ and ‘scene0000_00_vh_clean.segs.json’ in scene0000_00. Is the code reading both correct or just the v2 version.

**Comment by HalvesChen** [comment]
There is an error in the code I just gave you, try again.

```
coords, features, _, labels, inverse_map, pseudo_labels, inds, region, index, scenenames = data
region = region.squeeze()
raw_region = region.clone() # point region idx

in_field = ME.TensorField(features, coords, device=0)

if submodel is not None:
    feats_TensorField = model(in_field)
    feats_nonorm = submodel(feats_TensorField)
else:
    feats_nonorm = model(in_field)
feats_nonorm = feats_nonorm[inverse_map.long()] # point feats to voxel feats
feats_norm = F.normalize(feats_nonorm, dim=1) ## NOTE 可能需要normalize？没啥区别

region = region[inverse_map.long()] # point region idx to voxel region idx
region_feats = scatter(feats_norm, region.cuda(), dim=0, reduce='mean') # region feats

valid_mask = labels!=-1 # 获取带训练的mask区域
# labels = labels[valid_mask]
region_masked = region[valid_mask].long() ## NOTE
region_masked_num = torch.unique(region_masked) # get no-masked region
region_masked_feats = region_feats[region_masked_num] # get no-masked region feats
region_masked_feats_norm = F.normalize(region_masked_feats, dim=1).cpu() # norm

region_feats_list.append(region_masked_feats_norm)

torch.cuda.empty_cache()
torch.cuda.synchronize(torch.device("cuda"))
```

**Comment by HalvesChen** [comment]
> In the code for the initial sp scannet, I see a line that reads `parser.add_argument(‘--input_path’, type=str, default=‘data/ScanNet/scans/’, help=‘raw data path’) # *.segs .json file path`. Each of the `*.segs.json`s in the original dataset is one of two, e.g. ‘scene0000_00_vh_clean_2.0.010000.segs.json’ and ‘scene0000_00_vh_clean.segs.json’ in scene0000_00. Is the code reading both correct or just the v2 version.
Note this [line](https://github.com/SCUT-BIP-Lab/PointDC/blob/99f4f30ee4e2fa2d173c3881196afec4b02bb606/pointdc_mk/data_prepare/data_prepare_ScanNet.py#L24).

**Comment by HalvesChen** [comment]
Since I changed my workplace, it's a bit troublesome for me to debug it myself.

**Comment by David-Lmh** [comment]
I ran the updated code you provided last night, and the problem in the previous place was solved, but there was still the error reported mismatch in `eval_once` in eval_Scannet.py as follows

![7a92019bb222eec54683d7d8497aef4](https://github.com/user-attachments/assets/1e5fe05b-962d-4840-98b4-779090b828f8)

**Comment by HalvesChen** [comment]
This seems to be because the label maps from point to voxel, providing the following code.

**Comment by David-Lmh** [comment]
Ok, I will try it first. thank you very much

**Comment by David-Lmh** [comment]
I solved all the mismatches and the cause was incorrect initiation of superpoint. The reason for this is because in the `data_prepare/initialSP_Prepare_Scannet.py` file, line35 is `parser.add_argument(‘--input_path’, type=str, default='data/ScanNet/ scans/‘, help=’raw data path') # *.segs.json file path`. Because my dataset download contains both v1 and v2 versions and there is not enough storage space on the server. I copied `*.segs.json` separately, but in each scene I stored both versions of the segs.json file. This resulted in the superpoint being set incorrectly, which led to a mismatch between the number of regions (superpoints) and the number of features.

I re-did a separate v2 version of initial superpoint at a later time yesterday and have now successfully resolved the issue.

Thank you very much for your answer

**Comment by HalvesChen** [comment]
OK！
I want to confirm one thing: have you made any code changes? The changed codes https://github.com/SCUT-BIP-Lab/PointDC/issues/7#issuecomment-2587733768 is necessary?

**Comment by David-Lmh** [comment]
No code change. The changed codes https://github.com/SCUT-BIP-Lab/PointDC/issues/7#issuecomment-2587733768 is not necessary

**Comment by HalvesChen** [comment]
Got it 👍
